plugins/inputs/statsd/datadog.go

```diff
@@ -0,0 +1,187 @@
+	}
+
+	titleLen, err := strconv.ParseInt(rawLen[0], 10, 64)
+	if err != nil {
+		return fmt.Errorf("Invalid message format, could not parse title.length: '%s'", rawLen[0])
+	}
+
+	textLen, err := strconv.ParseInt(rawLen[1][:len(rawLen[1])-1], 10, 64)
+	if err != nil {
+		return fmt.Errorf("Invalid message format, could not parse text.length: '%s'", rawLen[0])
+	}
+	if titleLen+textLen+1 > int64(len(message)) {
+		return fmt.Errorf("Invalid message format, title.length and text.length exceed total message length")
+	}
+
+	rawTitle := message[:titleLen]
+	rawText := message[titleLen+1 : titleLen+1+textLen]
+	message = message[titleLen+1+textLen:]
+
+	if len(rawTitle) == 0 || len(rawText) == 0 {
+		return fmt.Errorf("Invalid event message format: empty 'title' or 'text' field")
+	}
+
+	// Handle hostname, with a priority to the h: field, then the host:
+	// tag and finally the defaultHostname value
+	// Metadata
+	m := event{
+		name: rawTitle,
+	}
+	m.tags = make(map[string]string, strings.Count(message, ",")+2) // allocate for the approximate number of tags
+	m.fields = make(map[string]interface{}, 9)
+	m.fields["alert_type"] = eventInfo // default event type
+	m.fields["text"] = uncommenter.Replace(string(rawText))
+	// host is a magic tag in the system, and it expects it to replace the result of h: if it is present
+	// telegraf will add a"host" tag anyway with different meaning than dogstatsd, so we need to use source instead of host.
+
+	m.tags["source"] = defaultHostname
+	m.fields["priority"] = priorityNormal
+	m.ts = now
+	if len(message) < 2 {
+		newM, err := tmetric.New(m.name, m.tags, m.fields, m.ts)
+		if err != nil {
+			return err
+		}
+		s.acc.AddMetric(newM)
+		return nil
+	}
+
+	rawMetadataFields := strings.Split(message[1:], "|")
+	for i := range rawMetadataFields {
+		if len(rawMetadataFields[i]) < 2 {
+			return errors.New("too short metadata field")
+		}
+		switch rawMetadataFields[i][:2] {
+		case "d:":
+			ts, err := strconv.ParseInt(rawMetadataFields[i][2:], 10, 64)
+			if err != nil {
+				continue
+			}
+			m.fields["ts"] = ts
+		case "p:":
+			switch rawMetadataFields[i][2:] {
+			case priorityLow:
+				m.fields["priority"] = priorityLow
+			case priorityNormal: // we already used this as a default
+			default:
+				continue
+			}
+		case "h:":
+			m.tags["source"] = rawMetadataFields[i][2:]
+		case "t:":
+			switch rawMetadataFields[i][2:] {
+			case eventError, eventWarning, eventSuccess, eventInfo:
+				m.fields["alert_type"] = rawMetadataFields[i][2:] // already set for info
+			default:
+				continue
+			}
+		case "k:":
+			m.tags["aggregation_key"] = rawMetadataFields[i][2:]
+		case "s:":
+			m.fields["source_type_name"] = rawMetadataFields[i][2:]
+		default:
+			if rawMetadataFields[i][0] == '#' {
+				parseDataDogTags(m.tags, rawMetadataFields[i][1:])
+			} else {
+				return fmt.Errorf("unknown metadata type: '%s'", rawMetadataFields[i])
+			}
+		}
+	}
+	if host, ok := m.tags["host"]; ok {
```

Comment:
How would m.tags have a host key?

Reply:
because in datadog's parser one way of setting the host (what we call source) is explicitly though a tag.  They treat that as equivalent to the h: value.